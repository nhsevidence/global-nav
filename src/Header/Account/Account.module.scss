@import '~@nice-digital/nds-core/scss/core';

.account {
  position: relative;
}

.button {
  background: #fff;
  border: 1px solid $colour-text;
  color: $colour-text;
  cursor: pointer;
  display: inline-block;
  font-family: inherit;
  font-size: inherit;
  font-weight: normal;
  height: rem(40); // TODO: Share a variable with search for height
  line-height: rem(40);
  padding: rem(0 $spacing-small);
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;

  &:focus {
    outline: 3px solid $colour-focus; // TODO: Replace with design system mixin
  }

  &:visited {
    color: $colour-text;
  }

  &:hover {
    border: 1px solid $colour-text; // NICE.bootstrap removes border on hover by default
  }
}

.myAccount {
  &:after {
    border-left: rem($spacing-x-small) solid transparent;
    border-right: rem($spacing-x-small) solid transparent;
    border-top: rem($spacing-small) solid $colour-nice-dark-grey;
    content: '';
    display: inline-block;
    height: 0;
    margin-left: rem($spacing-small);
    vertical-align: middle;
    width: 0;
  }

  &[aria-expanded='true']:after {
    border-bottom: rem($spacing-small) solid $colour-text;
    border-top: 0;
  }
}

.menu {
  background: $colour-true-white;
  border: 1px solid $colour-nice-dark-grey;
  list-style: none;
  margin: rem($spacing-small + 2px 0 0 0);
  padding: 0;
  position: absolute;
  right: 0;
  top: 100%;
  z-index: 2;

  // We move focus when we expand via keyboard navigation
  &:focus {
    // TODO: Use design system mixin
    outline: 3px solid $colour-focus;
  }

  &[aria-hidden='true'] {
    display: none;
  }

  &:before {
    border-bottom: rem($spacing-small) solid $colour-nice-dark-grey;
    border-left: rem($spacing-x-small) solid transparent;
    border-right: rem($spacing-x-small) solid transparent;
    bottom: 100%;
    content: '';
    display: block;
    height: 0;
    position: absolute;
    right: rem($spacing-medium);
    width: 0;
  }

  &:after {
    border-bottom: rem($spacing-small / 1.25) solid $colour-true-white;
    border-left: rem($spacing-x-small / 1.25) solid transparent;
    border-right: rem($spacing-x-small / 1.25) solid transparent;
    bottom: 100%;
    content: '';
    display: block;
    height: 0;
    margin-right: 1px;
    position: absolute;
    right: rem($spacing-medium);
    width: 0;
  }

  li {
    border-top: 1px solid $colour-nice-grey-2;
    list-style: none;
    margin: 0;
  }

  a {
    color: $colour-link;
    display: block;
    padding: rem($spacing-small $spacing-medium);
    text-decoration: none;
    white-space: nowrap;

    &:focus {
      // TODO: Use design system mixin
      outline: 3px solid $colour-focus;
    }

    &:hover {
      background: $colour-nice-grey-3;
      text-decoration: underline;
    }
  }
}
