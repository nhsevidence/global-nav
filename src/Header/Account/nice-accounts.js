// Checks if you are logged in via NICE Accounts
// Returns a promise that resolves with the data from NICE Accounts.
// Returns a promise that rejects if the data could not be loaded.
const checkIsLoggedIn = () => {
	return new Promise((resolve, reject) => {
		const url = "https://accounts.nice.org.uk/tophat";

		var body = document.body;
		var script = document.createElement("script");
		script.src =
			url +
			(~url.indexOf("?") ? "&" : "?") +
			Math.floor(Math.random() * 10000000000);

		let done = false;

		script.onload = script.onreadystatechange = () => {
			if (
				!done &&
				(!script.readyState ||
					script.readyState === "loaded" ||
					script.readyState === "complete")
			) {
				done = true;
				// Handle memory leak in IE
				script.onload = script.onreadystatechange = null;
				if (body && script.parentNode) {
					body.removeChild(script);
				}

				resolve(window._na);
			}
		};
		script.onerror = () => {
			reject();
		};

		body.insertBefore(script, body.firstChild);
	});
};

export { checkIsLoggedIn };
