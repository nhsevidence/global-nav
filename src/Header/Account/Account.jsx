import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import { checkIsLoggedIn } from "./nice-accounts";
import styles from "./Account.module.scss";

const escapeKeyCode = 27;

export default class Account extends Component {
	constructor(props) {
		super(props);

		this.state = {
			isExpanded: false
		};

		this.handleClick = this.handleClick.bind(this);
		this.handleKeyDown = this.handleKeyDown.bind(this);

		this.buttonRef = React.createRef();
		this.accountMenuRef = React.createRef();
	}

	handleClick(e) {
		const isKeyboardEvent = !e.pageX;
		this.setState(
			function(prevState) {
				return { isExpanded: !prevState.isExpanded };
			},
			function() {
				if (this.state.isExpanded && isKeyboardEvent) {
					this.accountMenuRef.current.setAttribute("tabIndex", -1);
					this.accountMenuRef.current.focus();
				}
			}.bind(this)
		);
	}

	handleKeyDown(event) {
		if (event.keyCode === escapeKeyCode) {
			event.preventDefault();
			this.setState({
				isExpanded: false
			});
			this.buttonRef.current.focus();
		}
	}

	componentDidMount() {
		checkIsLoggedIn()
			.then(
				function(data) {
					if (this.props.onLoginStatusChecked) {
						this.props.onLoginStatusChecked(data);
					}
				}.bind(this)
			)
			.catch(
				function(e) {
					console.warn("Couldn't load account data", e);
				}.bind(this)
			);
	}

	render() {
		const { accountsData } = this.props;

		return this.props.isLoggedIn ? (
			<div className={styles.account}>
				<button
					className={classnames(styles.button, styles.myAccount)}
					id="my-account-button"
					aria-controls="my-account"
					aria-haspopup="menu"
					aria-expanded={this.state.isExpanded}
					onClick={this.handleClick}
					onKeyDown={this.handleKeyDown}
					ref={this.buttonRef}
				>
					My account
				</button>
				<ul
					className={styles.menu}
					id="my-account"
					role="menu"
					aria-hidden={!this.state.isExpanded}
					aria-labelledby="my-account-button"
					ref={this.accountMenuRef}
				>
					{accountsData.links &&
						Object.keys(accountsData.links).map(
							function(text, i) {
								return (
									<li key={i} role="presentation">
										<a
											href={accountsData.links[text]}
											role="menuitem"
											onKeyDown={this.handleKeyDown}
											data-hj-suppress={
												accountsData.links[text].indexOf("profile") > -1
													? ""
													: null
											}
										>
											{text}
										</a>
									</li>
								);
							}.bind(this)
						)}
				</ul>
			</div>
		) : (
			<a href="https://accounts.nice.org.uk/signin" className={styles.button}>
				Sign in
			</a>
		);
	}
}

Account.propTypes = {
	isLoggedIn: PropTypes.bool.isRequired,
	onLoginStatusChecked: PropTypes.func,
	accountsData: PropTypes.shape({
		display_name: PropTypes.string,
		thumbnail: PropTypes.string,
		links: PropTypes.object
	})
};
