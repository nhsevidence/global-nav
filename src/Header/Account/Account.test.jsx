import React from "react";
import Account from "./Account";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { checkIsLoggedIn } from "./nice-accounts";

jest.mock("./nice-accounts", () => ({
	checkIsLoggedIn: jest.fn(() =>
		Promise.resolve({
			test: true
		})
	)
}));

describe("Account", () => {
	const accountsData = {
		display_name: "Joe Bloggs",
		links: {
			"Joe Bloggs": "https://accounts.nice.org.uk/users/12345/editprofile",
			"Sign out": "https://accounts.nice.org.uk/signout"
		}
	};

	it("Renders without crashing", () => {
		const wrapper = shallow(<Account isLoggedIn={false} />);
		expect(wrapper).toHaveLength(1);
	});

	it("Matches snapshot when logged out", () => {
		const wrapper = shallow(<Account isLoggedIn={false} />);

		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Matches snapshot when logged in", () => {
		const wrapper = shallow(
			<Account isLoggedIn={true} accountsData={accountsData} />
		);

		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Calls onLoginStatusChecked callback prop when mounted", done => {
		const onLoginStatusChecked = jest.fn();

		//checkIsLoggedIn = jest.fn(() => Promise.resolve());

		const wrapper = shallow(
			<Account isLoggedIn={false} onLoginStatusChecked={onLoginStatusChecked} />
		);

		setImmediate(() => {
			expect(onLoginStatusChecked).toHaveBeenCalledWith({ test: true });
			done();
		});
	});

	it("Expands account menu when button is clicked", () => {
		const wrapper = shallow(
			<Account isLoggedIn={true} accountsData={accountsData} />
		);

		wrapper.find("[aria-controls='my-account']").simulate("click");

		expect(
			wrapper.find("[aria-controls='my-account']").props()["aria-expanded"]
		).toBe(true);
		expect(wrapper.find("#my-account").props()["aria-hidden"]).toBe(false);
	});
});
