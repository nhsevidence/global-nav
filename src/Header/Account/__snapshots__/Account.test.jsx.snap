// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Matches snapshot when logged in 1`] = `
<div
  className="account"
>
  <button
    aria-controls="my-account"
    aria-expanded={false}
    className="button myAccount"
    onClick={[Function]}
  >
    My account
  </button>
  <ul
    aria-hidden={true}
    className="menu"
    id="my-account"
  >
    <li
      key="0"
    >
      <a
        data-hj-suppress=""
        href="https://accounts.nice.org.uk/users/12345/editprofile"
      >
        Joe Bloggs
      </a>
    </li>
    <li
      key="1"
    >
      <a
        data-hj-suppress={null}
        href="https://accounts.nice.org.uk/signout"
      >
        Sign out
      </a>
    </li>
  </ul>
</div>
`;

exports[`Account Matches snapshot when logged out 1`] = `
<a
  className="button"
  href="https://accounts.nice.org.uk/signin"
>
  Sign in
</a>
`;
